### PR TITLE
D7YIGPositionCalibration new fitting approach

### PR DESCRIPTION
**Pull request by darominski**
**Description of work.**

Added the option to choose between fitting the YIG Bragg peaks individually or all of peaks recorded by a single detector simultaneously. Part of this new functionality is an option to **not** perform fitting but provide feedback about the initial guesses to the user by plotting the Gaussian peaks overlaid on the loaded data. 

**To test:**

Run the D7 YIG position calibration using D7 UnitTest data setting `FittingMethod` property to 'None', 'Individual', and 'Global' using the example below. 

```
Load('ILL/D7/396442_396831.nxs', OutputWorkspace='intermediateWavelengthScan')
approximate_wavelength = '4.8' # Angstrom
D7YIGPositionCalibration(InputWorkspace='intermediateWavelengthScan', ApproximateWavelength=approximate_wavelength,
                                         YIGPeaksFile='D7_YIG_peaks.xml', CalibrationOutputFile='test_intermediateWavelength.xml',
                                         MinimalDistanceBetweenPeaks=1.5, BankOffsets="-4,-4,0",
                                         FittingMethod='None', ClearCache=False,
                                         FitOutputWorkspace='intermediateWavelength')
```
1. See whether the overlaid Gaussian peaks guesses are close to peaks in data.

2. With FittingMethod either 'Individual', and 'Global', set the LogLevel to notice to see the output parameters of the calibration: wavelength, and three bank gradients. The wavelength should be close close to 4.86 and the gradients should be consistent with unity down to 1%.

*There is no associated issue.*

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Review by Harrietbrown — COMMENTED** [review]
Running with `Individual` is inconsistent, sometimes it sucseeds, others I get an error. It also takes a long time with both Individual and global.
```Error in execution of algorithm D7YIGPositionCalibration:
list index out of range
  at line 169 in 'C:/Users/wey38795/Documents/mantid_builds/Master/raw/mantid/Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/D7YIGPositionCalibration.py'
  caused by line 548 in 'C:/Users/wey38795/Documents/mantid_builds/Master/raw/mantid/Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/D7YIGPositionCalibration.py'
RuntimeError: list index out of range
  at line 169 in 'C:/Users/wey38795/Documents/mantid_builds/Master/raw/mantid/Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/D7YIGPositionCalibration.py'
  caused by line 548 in 'C:/Users/wey38795/Documents/mantid_builds/Master/raw/mantid/Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/D7YIGPositionCalibration.py'
  File "<string>", line 8, in <module>
  File "c:\users\wey38795\documents\mantid_builds\master\raw\mantid\framework\pythoninterface\mantid\simpleapi.py", line 1036, in __call__
    algm.execute()```

File: docs/source/algorithms/D7YIGPositionCalibration-v1.rst [hunk 1]
```diff
@@ -60,10 +62,11 @@ Usage
    Load('ILL/D7/396442_396831.nxs', OutputWorkspace='intermediateWavelengthScan')
    approximate_wavelength = '4.8' # Angstrom
    D7YIGPositionCalibration(InputWorkspace='intermediateWavelengthScan', ApproximateWavelength=approximate_wavelength,
-                               YIGPeaksFile='D7_YIG_peaks.xml', CalibrationOutputFile='test_intermediateWavelength.xml',
-                               MinimalDistanceBetweenPeaks=1.5, BankOffsets="-4,-4,0",
-                               FittingMethod='Individual', FitOutputWorkspace='intermediateWavelength')
-		       
+                            YIGPeaksFile='D7_YIG_peaks.xml', CalibrationOutputFile='test_intermediateWavelength.xml',
+                            MinimalDistanceBetweenPeaks=1.5, BankOffsets=[-4,-4,0],
+			    MaskedBinsRange=[-50, -25, 15],
```
Comment (Harrietbrown) [inline docs/source/algorithms/D7YIGPositionCalibration-v1.rst:67]: Make sure these ahve the right indentation.

**Review by Harrietbrown — CHANGES_REQUESTED** [review]
One minor document formatting issue, other than that the script now runs in a reasonable time and does not produce an error. and the calibration values are as described.

**Comment by gvardany** [comment]
Could you also add a test for the Global method?

**Comment by darominski** [comment]
> Could you also add a test for the Global method?

Good point, naturally it should be added. Also, preferably as part of the already existing system test, instead of the unit test. Otherwise, the test time increases from the current ~10 s to about 20 s, which is probably too excessive.

**Comment by darominski** [comment]
> Running with `Individual` is inconsistent, sometimes it sucseeds, others I get an error. It also takes a long time with both Individual and global.
> 
> ```
> list index out of range
>   at line 169 in 'C:/Users/wey38795/Documents/mantid_builds/Master/raw/mantid/Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/D7YIGPositionCalibration.py'
>   caused by line 548 in 'C:/Users/wey38795/Documents/mantid_builds/Master/raw/mantid/Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/D7YIGPositionCalibration.py'
> RuntimeError: list index out of range
>   at line 169 in 'C:/Users/wey38795/Documents/mantid_builds/Master/raw/mantid/Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/D7YIGPositionCalibration.py'
>   caused by line 548 in 'C:/Users/wey38795/Documents/mantid_builds/Master/raw/mantid/Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/D7YIGPositionCalibration.py'
>   File "<string>", line 8, in <module>
>   File "c:\users\wey38795\documents\mantid_builds\master\raw\mantid\framework\pythoninterface\mantid\simpleapi.py", line 1036, in __call__
>     algm.execute()```
> ```

Indeed, there was an issue with running the algorithm multiple times without changing the parameters names and with `ClearCache` property set to `False` due to the way conjoining workspaces was used. This has been corrected by checking whether a workspace with such name exists in the ADS and removing it if that's the case.

File: docs/source/algorithms/D7YIGPositionCalibration-v1.rst (hunk 1 above)
Comment (darominski) [reply docs/source/algorithms/D7YIGPositionCalibration-v1.rst:67]: Well spotted. This is a little bit strange, as they are all aligned on my local machine and were edited with emacs. When I open this through gedit, they also look properly aligned. I am going to correct it from github for now but need to find a source of this issue.

Comment (darominski) [reply docs/source/algorithms/D7YIGPositionCalibration-v1.rst:67]: i agree it looks like there's a wrong indentation, however when I open this file in emacs, or view or edit it on github, these lines seem properly aligned and indented. That's a bit strange

Comment (Harrietbrown) [reply docs/source/algorithms/D7YIGPositionCalibration-v1.rst:67]: You were using tabs for some lines and spaces for others.

Comment (darominski) [reply docs/source/algorithms/D7YIGPositionCalibration-v1.rst:67]: Oh, right. The most recent commit should've fixed the issue.

**Review by Harrietbrown — APPROVED** [review]
That's got the rst sorted, and the global and Individual versions are much faster. Good job.

**Pull request merged**